### DISCUSSION TITLE
no cfs in rviz when using lps

### DISCUSSION THREAD

**Discussion by Supredan** [USC-ACTLab/crazyswarm#339] [Q&A]
Hi,  
I am using LPS to run crazyswarm, everything works fine, except that rviz cannot display the cf's trace.
but [here](https://crazyswarm.readthedocs.io/en/latest/tutorials/tutorials.html) says :
> Furthermore, rviz will show the estimated pose of all CFs.

i also see [here](https://github.com/USC-ACTLab/crazyswarm/issues/108#issuecomment-432350943)  #108 :

> Yes - there is no data streaming. One possibility would be to enable logging, log the current state estimate, and sent this to rviz. This hasn't been implemented so far (and you would need a good number of radios for a high update rate.)

But I still can’t subscribe any useful topic in rviz after Enable the log. Is there any more detailed instructions or tutorials? Thanks very much!!!

**Comment by Supredan** [comment]
```
<?xml version="1.0"?>
<launch>
  <arg name="joy_dev" default="/dev/input/js0" />

  <rosparam command="load" file="$(find crazyswarm)/launch/crazyflieTypes.yaml" />
  <rosparam command="load" file="$(find crazyswarm)/launch/crazyflies.yaml" />

  <node pkg="crazyswarm" type="crazyswarm_server" name="crazyswarm_server" output="screen" >
    <rosparam>
      world_frame: "/world"
      # Logging configuration (Use enable_logging to actually enable logging)
      genericLogTopics: ["log1"]
      genericLogTopicFrequencies: [10]
      genericLogTopic_log1_Variables: ["stateEstimate.z", "ctrltarget.x"]
      # firmware parameters for all drones (use crazyflieTypes.yaml to set per type, or
      # allCrazyflies.yaml to set per drone)
      firmwareParams:
        commander:
          enHighLevel: 1
        stabilizer:
          estimator: 2 # 1: complementary, 2: kalman
          controller: 2 # 1: PID, 2: mellinger
        ring:
          effect: 16 # 6: double spinner, 7: solid color, 16: packetRate
          solidBlue: 255 # if set to solid color
          solidGreen: 0 # if set to solid color
          solidRed: 0 # if set to solid color
          headlightEnable: 0
        locSrv:
          extPosStdDev: 1e-3
          extQuatStdDev: 0.5e-1
        kalman:
          resetEstimation: 1
      # tracking
      motion_capture_type: "none" # one of none,vicon,optitrack,qualisys,vrpn
      object_tracking_type: "libobjecttracker" # one of motionCapture,libobjecttracker
      send_position_only: False # set to False to send position+orientation; set to True to send position only
      vicon_host_name: "vicon"
      # optitrack_host_name: "optitrack"
      # qualisys_host_name: "10.0.5.219"
      # qualisys_base_port: 22222
      # vrpn_host_name: "vicon"
      save_point_clouds: ~/pointCloud.ot
      print_latency: False
      write_csvs: False
      force_no_cache: False
      enable_parameters: True
      enable_logging: True
    </rosparam>
  </node>

  <node name="joy" pkg="joy" type="joy_node" output="screen">
    <param name="dev" value="$(arg joy_dev)" />
  </node>

  <node pkg="crazyswarm" type="crazyswarm_teleop" name="crazyswarm_teleop" output="screen">
    <param name="csv_file" value="$(find crazyswarm)/launch/figure8_smooth.csv" />
    <param name="timescale" value="0.8" />
  </node>

  <node name="rviz" pkg="rviz" type="rviz" args="-d $(find crazyswarm)/launch/test.rviz"/>

  <!-- <node pkg="rqt_plot" type="rqt_plot" name="rqt_plot_x" args="/cf2/log1/values[0]"/> -->
  <!-- <node pkg="rqt_plot" type="rqt_plot" name="rqt_plot_roll" args="/cf1/log1/values[2] /cf1/log1/values[3]"/> -->

</launch>
```

BTW  here is my hover_swarm.launch, should i make some change in **Logging configuration**?

**Comment by jpreiss** [comment]
RViz only shows the poses when you are using a motion capture system. With LPS, the PC doesn't know the positions.

There is no built-in functionality in Crazyswarm to convert the logging variables into ROS transforms that RViz will show. You would need to write a ROS node that converts the logging topics (like `cf2/log1/values[0]`) into ROS transforms, and add it to your launch file.

If you write such a utility, we would appreciate it if you contribute the code back to Crazyswarm in a PR for the benefit of other users!

Like Wolfgang mentioned in #108, this will use a lot of radio bandwidth.

For info on configuring the logging system, check pp31-32 of [Flying Multiple UAVs Using ROS](act.usc.edu/publications/Hoenig_Springer_ROS2017.pdf) (pdf). It would be good to add this to Crazyswarm's docs in the future.

**Reply by Supredan** [reply]
Thank you for your help!
I wrote a simple Toy ros publisher as you said, and successfully saw the cf trace in rviz. Although there was some delay, it worked.
I will submit the code after finishing it. Thanks again!

**Reply by Supredan** [reply]
Sorry to bother again, but I have a problem when testing multiple cfs -_-!

when i test 2 cfs everything goes fine and two traces shows in rviz. But when the number is increased to 3, only the last two cf logdata can be received( i use **[rostopic echo]** to see the message), I also increased the number to 6, still only the last 2 cf display data.

I tried to reduce the log frequency from 10ms to 100ms, but the problem still exists, so the problem should not be the bandwidth. Is there a place where the maximum number of cf is configured? Thanks in advance!!!

**Reply by whoenig** [reply]
You can just set enable_logging_pose to true, and you should get a /cf<id>/pose topic that you can visualize in rviz. See https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/src/crazyswarm_server.cpp#L201. This updates every 10 Hz by default.

**Reply by jpreiss** [reply]
Apologies @Supredan, I forgot about that feature even though I was the code reviewer. Hope I did not waste too much of your time with my initial suggestion.

**Reply by jpreiss** [reply]
@whoenig this means we can close #145, correct?

**Reply by WaterHyacinthInNANHU** [reply]
> You can just set enable_logging_pose to true, and you should get a /cf/pose topic that you can visualize in rviz. See https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/src/crazyswarm_server.cpp#L201. This updates every 10 Hz by default.

Thanks for your useful suggestion!
I have the same problem here when I enable 3 or more cfs, no more than 3 `/cfx/log1` topics are logging. Sometimes none of  `/cfx/log1` topics is logging at all.
Also, I've tried to set `enable_logging_pose` to `true` and pose data only shows up on topic `/cfx/pose` when `/cfx/log1` have data being logged. I guess there's a bandwidth limitation here?

So my quesitons are:
1) I understand if more than 2 cfs are logging the communication may jam, but how could I decrease the transmitting rate to support more logging cfs?
2) I'm confused that under lps, where's pose been estimated? Is it estimated on-board and then transmitted to the ground station? If not then what data exactly is transmitting between them under lps?

**Reply by whoenig** [reply]
This sounds like a potential bug. I have not tried that myself, as I usually disable logging when flying more than two CFs.
1. You can reduce the update rate of the logging variables.
2. The pose is estimated using the on-board Kalman filter. By default, there is no transmission to the ground station. If you enable the /cf<id>/pose topic, the logging variables contain the output of the Kalman filter that runs on-board.

**Reply by WaterHyacinthInNANHU** [reply]
Thanks for your help!

For now, out of my understanding, cfs only transmit their logging variables when I set `enable_logging` to `True`. Otherwise, there's no data been transmitted to the ground station under lps. Am I correct in this regard?

Few more questions about this topic:
1) Where should I modify to reduce the logging rate? Should I modify the crazyfies firmware?
2) [Here](https://crazyswarm.readthedocs.io/en/latest/configuration.html#prepare-crazyflies) says:
> Each radio can control about 15 Crazyflies. If you have more than 15 CFs you will need to assign different channels to the Crazyflies. For example, if you have 49 Crazyflies you’ll need three unique channels. It is up to you which channels you assign to which CF, but a good way is to use the Crazyflie number modulo the number of channels. For example, cf1 is assigned to channel 80, cf2 is assigned to channel 90, cf3 is assigned to channel 100, cf4 is assigned to channel 80 and so on.

I do not quite understand the term `channel` here. If it stands for a specific transmitting frequency band, is it feasible to use more `channel` to expand the number of logging cfs? For example, 3 channel `80, 90, 100` to support at most 6 logging cfs?

**Reply by whoenig** [reply]
0. Yes, you are correct that enable_logging needs to be set to True.
1. For custom log topics [here](https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/launch/hover_swarm.launch#L13). For the pose, you need to change the source code [here](https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/src/crazyswarm_server.cpp#L662). You might also be able to support more CFs per radio by increasing the value [here](https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/src/crazyswarm_server.cpp#L1016).
2. Yes, but each Crazyflie is bound to a channel and each radio is serving exactly one channel. So if you use more channels, you will need more radios and each radio will server fewer Crazyflies.

**Reply by sabdi1** [reply]
@whoenig 

> You can just set enable_logging_pose to true, and you should get a /cf/pose topic that you can visualize in rviz. See https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/src/crazyswarm_server.cpp#L201. This updates every 10 Hz by default.

As mentioned here, I set the enable_logging_pose to True and enable_logging to False, but I could not find any cfx/pose/ topic being published. If I set both enable_logging_pose and enable_logging to True, I get an error saying 

what():  Could not find stateEstimate.x in log toc!

Do I set both of them to true? Am I missing something here?

**Reply by whoenig** [reply]
Both need to be enabled.

**Reply by WaterHyacinthInNANHU** [reply]
> 0. Yes, you are correct that enable_logging needs to be set to True.
> 1. For custom log topics [here](https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/launch/hover_swarm.launch#L13). For the pose, you need to change the source code [here](https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/src/crazyswarm_server.cpp#L662). You might also be able to support more CFs per radio by increasing the value [here](https://github.com/USC-ACTLab/crazyswarm/blob/master/ros_ws/src/crazyswarm/src/crazyswarm_server.cpp#L1016).
> 2. Yes, but each Crazyflie is bound to a channel and each radio is serving exactly one channel. So if you use more channels, you will need more radios and each radio will server fewer Crazyflies.

Sorry for getting back to you late.Thanks for your reply!
I tested with 3 PAs with 2 crazyflies logging to each PA and it works just fine!

**Reply by sabdi1** [reply]
> Both need to be enabled.

I'm able to get the pose information with LPS from /cf1/pose/ topic and pose is correct. Now I have replaced the LPS deck with flow deck. The position information in /cf1/pose/ topic doesn't seem to be right. Should I be doing something different for flow deck?

**Reply by whoenig** [reply]
No, it's the same approach. Since the flow deck can only directly measure velocity, the position will drift over time.
